### PR TITLE
In main loop, add `on_stats_available` callback

### PR DESCRIPTION
**Pull request by akavel**
### Pull Request Description

This change adds an `on_stats_available` callback hook in the main loop of the `World` type. Handlers registered on this callback will be called when `StatsData` of the most recent frame is fully calculated and available for use (the `StatsData` is passed as an argument to the called handlers).

This callback hook is intended to be used in the future by a profiling framework.

### Visuals

The demo scene `stats` was updated to make use of the new code.

https://user-images.githubusercontent.com/273837/153028305-7bafd00f-50f3-4fbb-819f-de2a32b9bb97.mov

A screenshot of the Monitor panel in full `ide` after applying the PR, taken in IDE built with `./run dist`:

Screenshot 2022-02-08 at 17 35 11

https://www.pivotaltracker.com/story/show/181093832

### Important Notes

 - Note: the `Monitor` panel's code performs an optimization, such that it skips `Stats::begin_frame()` and `Stats::end_frame()` calls if the panel hidden from view. The code was updated to make sure that any frames where `Monitor` was hidden are not being included in calculations, to avoid emitting invalid stats values.
 - An unused `PanelData::begin_value` field was removed.

[ci no changelog needed]

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/develop/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/develop/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: lib/rust/ensogl/core/src/display/world/stats.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/world/stats.rs]: Why function that "Returns a snapshot of statistics data for the previous rendering frame ..." is called "begin"? For me that does not make sense now and I do not understand the behavior/semantics of this function anymore.

File: lib/rust/ensogl/core/src/display/world/stats.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/world/stats.rs]: I'm missing info why this is tracked and what does this information influence. I was trying to read that from the other changes, but then I have seen that function "begin" does something strange and I'm lost right now in understanding of the logic here.

**Review by wdanilo — CHANGES_REQUESTED** [review]

**Comment by akavel** [comment]
@wdanilo I did some refactoring inspired by your comments; would you possibly find the code clearer now?

**Review by wdanilo — APPROVED** [review]
Yup, it's very clear now!

**Comment by akavel** [comment]
(*Process note:* the code-review is now accepted, therefore the PR is waiting for Review by @MichaelMauderer on Task https://www.pivotaltracker.com/story/show/181093832 before merge.)

**Pull request merged**
